pkg/server/service/loadbalancer/wrr/wrr.go

```diff
@@ -37,79 +39,66 @@ func New(sticky *dynamic.Sticky) *Balancer {
 	return balancer
 }
 
-// Balancer is a WeightedRoundRobin load balancer.
-type Balancer struct {
-	handlers []*namedHandler
-	mutex    *sync.Mutex
-	// Current index (starts from -1)
-	index         int
-	currentWeight int
-	stickyCookie  *stickyCookie
+type namedHandlers []*namedHandler
+
+// Len implements heap.Interface/sort.Interface
+func (n namedHandlers) Len() int { return len(n) }
+
+// Less implements heap.Interface/sort.Interface
+func (n namedHandlers) Less(i, j int) bool {
+	return n[i].deadline < n[j].deadline
 }
 
-func (b *Balancer) maxWeight() int {
-	max := -1
-	for _, s := range b.handlers {
-		if s.weight > max {
-			max = s.weight
-		}
-	}
-	return max
+// Swap implements heap.Interface/sort.Interface
+func (n namedHandlers) Swap(i, j int) {
+	n[i], n[j] = n[j], n[i]
 }
 
-func (b *Balancer) weightGcd() int {
-	divisor := -1
-	for _, s := range b.handlers {
-		if divisor == -1 {
-			divisor = s.weight
-		} else {
-			divisor = gcd(divisor, s.weight)
-		}
+// Push implements heap.Interface for pushing an item into the heap
+func (n *namedHandlers) Push(x interface{}) {
+	h, ok := x.(*namedHandler)
+	if !ok {
+		return
 	}
-	return divisor
+	*n = append(*n, h)
 }
 
-func gcd(a, b int) int {
-	for b != 0 {
-		a, b = b, a%b
-	}
-	return a
+// Pop implements heap.Interface for poping an item from the heap
+func (n *namedHandlers) Pop() interface{} {
+	old := *n
+	h := old[len(old)-1]
+	*n = old[0 : len(old)-1]
+	return h
+}
+
+// Balancer is a WeightedRoundRobin load balancer based on Earliest Deadline First (EDF).
+// (https://en.wikipedia.org/wiki/Earliest_deadline_first_scheduling)
+// Each pick from the schedule has the earliest deadline entry selected. Entries have deadlines set
+// at currentDeadline + 1 / weight, providing weighted round robin behavior with floating point
+// weights and an O(log n) pick time.
+type Balancer struct {
+	stickyCookie *stickyCookie
+
+	mutex       sync.RWMutex
+	handlers    *namedHandlers
```

Comment:
I think this field does not have to be a pointer, and everywhere needed we could call the methods on &b.handlers instead of b.handlers, but I don't have any strong argument in favor of that for now, so I'll defer to the other reviewers.

Reply:
After consulting with co-workers, here is a reason why we should try to do better: if we move the implementation of Pop (and others) one level up, i.e. to the Balancer instead of directly on the handlers, then we might be able to remove the extra indirection (handlers would become a slice again instead of a pointer), which would remove a bit of complexity. Can you please try to see if it can be done?

Reply:
Solved.

Reply:
Perfect, it indeed all looks much better now, thanks.